### PR TITLE
Fix nearest minute rounding

### PR DESCRIPTION
**Pull request by MelvinVermeer**
Fixes: #3195 & #3268

The issue that I was facing was that when rewrote this test:

```ts
it('rounds according to the passed mode - floor - when nearestTo is provided', () => {
  const result = roundToNearestMinutes(
    new Date(2014, 6 /* Jul */, 10, 12, 10, 30, 5),
    { nearestTo: 10, roundingMethod: 'floor' }
  )
  assert.deepStrictEqual(result, new Date(2014, 6 /* Jul */, 10, 12, 10, 0))
})
```

to this

```diff
it('rounds according to the passed mode - floor - when nearestTo is provided', () => {
  const result = roundToNearestMinutes(
-    new Date(2014, 6 /* Jul */, 10, 12, 10, 30, 5),
+    new Date(2014, 6 /* Jul */, 10, 12, 19, 30, 5),

    { nearestTo: 10, roundingMethod: 'floor' }
  )
  assert.deepStrictEqual(result, new Date(2014, 6 /* Jul */, 10, 12, 10, 0))
})
```

My expectation was that `12:19:30:500` would be rounded down (floor) to the nearest interval on 10 minutes, i.e. `12:10:00:000`

This updated test failed.

I have updated the tests and implementation accordingly.

### CODE REVIEW

File: src/roundToNearestMinutes/test.ts [hunk 1]
```diff
@@ -70,7 +71,7 @@ describe('roundToNearestMinutes', () => {
       new Date(2014, 6 /* Jul */, 10, 12, 10, 30, 5),
       { roundingMethod: 'ceil' }
     )
-    assert.deepStrictEqual(result, new Date(2014, 6 /* Jul */, 10, 12, 12, 0))
+    assert.deepStrictEqual(result, new Date(2014, 6 /* Jul */, 10, 12, 11, 0))
```
Comment (MelvinVermeer) [inline src/roundToNearestMinutes/test.ts:74]: The original expectation rounds the time up from `12:10:30:5`, to `12:12:00:000` - a false assertion i guess

File: src/roundToNearestMinutes/test.ts [hunk 2]
```diff
@@ -54,7 +55,7 @@ describe('roundToNearestMinutes', () => {
       new Date(2014, 6 /* Jul */, 10, 12, 10, 30, 5),
       { roundingMethod: 'floor' }
     )
-    assert.deepStrictEqual(result, new Date(2014, 6 /* Jul */, 10, 12, 11, 0))
+    assert.deepStrictEqual(result, new Date(2014, 6 /* Jul */, 10, 12, 10, 0))
```
Comment (MelvinVermeer) [inline src/roundToNearestMinutes/test.ts:58]: Incorrect assertion if i pass in a rounding method 'floor' i expect to round down from `12:10:30` to `12:10:00` and not `12:11:00`

File: src/roundToNearestMinutes/test.ts [hunk 3]
```diff
@@ -21,14 +21,15 @@ describe('roundToNearestMinutes', () => {
   it('rounds to the closest x minutes if nearestTo is provided', () => {
     const result = roundToNearestMinutes(
       new Date(2014, 6 /* Jul */, 10, 12, 10, 30),
-      { nearestTo: 4 }
+      { nearestTo: 4, roundingMethod: 'round' }
```
Comment (MelvinVermeer) [inline src/roundToNearestMinutes/test.ts:24]: the assertion in the test is based on 'normal' rounding behaviour, while `trunc` is date-fns' default.

File: src/roundToNearestMinutes/test.ts [hunk 4]
```diff
@@ -21,14 +21,15 @@ describe('roundToNearestMinutes', () => {
...
     )
     assert.deepStrictEqual(result, new Date(2014, 6 /* Jul */, 10, 12, 12, 0))
   })
 
   it('rounds up >=30 seconds for nearestTo=1', () => {
     const result = roundToNearestMinutes(
-      new Date(2014, 6 /* Jul */, 10, 12, 13, 30)
+      new Date(2014, 6 /* Jul */, 10, 12, 13, 30),
+      { roundingMethod: 'round' }
```
Comment (MelvinVermeer) [inline src/roundToNearestMinutes/test.ts:32]: the assertion in the test is based on 'normal' rounding behaviour, while trunc is date-fns' default.

File: src/roundToNearestMinutes/test.ts (hunk 3 above)
Comment (MelvinVermeer) [reply src/roundToNearestMinutes/test.ts:24]: other option is to update the assertion

File: src/roundToNearestMinutes/test.ts (hunk 4 above)
Comment (MelvinVermeer) [reply src/roundToNearestMinutes/test.ts:32]: other option is to update the assertion

**Review by TimNuman — APPROVED** [review]

**Comment by MelvinVermeer** [comment]
I understand that this change is considered breaking, and might belong in a major update only.

**Comment by Bysmyyr** [comment]
> I understand that this change is considered breaking, and might belong in a major update only.

Could it be added as a new roundingMethod for now so it does not change anything and those who want to use it could use it. Something like, "fixedFloor"

**Comment by vladshcherbin** [comment]
Such basic feature is working incorrectly when the fix's ready 🤦

**Comment by mp035** [comment]
There seems to be no mention of it here, but roundToNearestMinute has  a bug which will be fixed when [pr 3467](https://github.com/date-fns/date-fns/pull/3467) is merged.  My cursory reading of this issue indicates that the bug may be clouding understanding of the underlying issue.

**Comment by johnhunter** [comment]
This is a change in behaviour but I wonder how many are using it since it doesn't work as expected when the round method is "floor".

Would be nice to see this released.

File: src/roundToNearestMinutes/index.ts [hunk 5]
```diff
@@ -46,14 +46,13 @@ export function roundToNearestMinutes<DateType extends Date>(
   if (nearestTo < 1 || nearestTo > 30) return constructFrom(date, NaN);
 
   const _date = toDate(date);
-  const seconds = _date.getSeconds(); // relevant if nearestTo is 1, which is the default case
-  const minutes = _date.getMinutes() + seconds / 60;
+  const seconds = date.getSeconds(); // relevant if nearestTo is 1, which is the default case
```
Comment (kossnocorp) [inline src/roundToNearestMinutes/index.ts:49]:
```suggestion
  const seconds = _date.getSeconds(); // relevant if nearestTo is 1, which is the default case
```

File: src/roundToNearestMinutes/index.ts [hunk 6]
```diff
@@ -47,13 +47,12 @@ export function roundToNearestMinutes<DateType extends Date>(
 
   const _date = toDate(date);
   const seconds = _date.getSeconds(); // relevant if nearestTo is 1, which is the default case
-  const minutes = _date.getMinutes() + seconds / 60;
+  const minutes = date.getMinutes() + seconds / 60;
```
Comment (kossnocorp) [inline src/roundToNearestMinutes/index.ts:50]:
```suggestion
  const minutes = _date.getMinutes() + seconds / 60;
```

File: src/roundToNearestMinutes/index.ts [hunk 7]
```diff
@@ -47,13 +47,12 @@ export function roundToNearestMinutes<DateType extends Date>(
...
   const roundingMethod = getRoundingMethod(options?.roundingMethod);
+
   const roundedMinutes = roundingMethod(minutes / nearestTo) * nearestTo;
-  const remainderMinutes = minutes % nearestTo;
-  const addedMinutes = Math.round(remainderMinutes / nearestTo) * nearestTo;
+  const result = constructFrom(date, date);
```
Comment (kossnocorp) [inline src/roundToNearestMinutes/index.ts:54]:
```suggestion
  const result = constructFrom(date, _date);
```

**Comment by fturmel** [comment]
@kossnocorp please consult #3132 while working on this. It should be useful even though it was based on v2. It has fixes for bad unit tests, fractional milliseconds bug, docs improvements, etc.

**Comment by kossnocorp** [comment]
@fturmel, thank you for the heads up. I'm on it! I want to compare the fix implementations, as I'm not very familiar with the function.

**Comment by kossnocorp** [comment]
Thank you so much for your work. I'll make sure to mention you in the release notes. I used this PR as it covers more cases, adds more edge case tests and was also proposed a bit earlier: https://github.com/date-fns/date-fns/pull/3132

**Comment by kossnocorp** [comment]
Shipped with https://github.com/date-fns/date-fns/releases/tag/v3.2.0
